### ISSUE TITLE
Is it easy to write scripts to call MTGA client API?

### ISSUE THREAD

**Issue by iscas-zac** [MayerDaniel/mayerdaniel.github.io#1]
Hello, and sorry to bother you in the github, because I think the guestbook is not for your replying.

I'm impressed by the articles, and wondering if you have done all the reversing work alone and have more information about MTGA client APIs. The client responds too slowly in my machine, and I'm thinking of manipulating it with some scripts.

This is only a thought, and I'm assessing if it will truly save my time to hack on such things (for example make moves without closing the annoying "reconnect" window or waiting for the two avatar loading scene) with scripts. Can I do that easily?

**Comment by MayerDaniel** [comment]
Yeah, check out the code at the bottom of https://www.mayer.cool/writings/I-Hacked-Magic-the-Gathering/ for code that instantiates a copy of sparky that you can mess around with. That bot object has a fully functioning GRE client, which is essentially the bare-bones magic engine that holds the game state that you interact with when playing. You can then create a text-only interface if you so choose, but that would only help you if graphics were the problem. The network communication will be the exact same.

**Comment by iscas-zac** [comment]
Thanks for replying, I'll give it a try.

**Comment by iscas-zac** [comment]
Hello, I’ve been hacking on MTGA’s network protocol. I downloaded dnSpyEx, and it’s of great help. 

But the call stack is still complicated, and I wonder how you found out that Sparky is a local AI only by inspecting the code, while bot draft (as you didn’t mention) is probably not.

 I’d love to know if you play with some other tools or methods good at this, or other people and online information about MTGA’s protocols (like [riQQ’s protocol](https://github.com/riQQ/MtgaProto)). If you are interested, I am making a script to mock the client and am happy to share it.

**Comment by MayerDaniel** [comment]
Yeah put it on github! That sounds pretty interesting. I haven't looked into the network protocol that much, but something that might help you is the built-in debug tools. Checkout the function hasrole_debugging or something like that. If you patch that to return true, when you press alt you will see a debugging overlay that has a tab all about what messages are being sent and received

**Comment by iscas-zac** [comment]
I've put it in [a repo](https://github.com/duelist-X/MTGate). It's basically very naive, but contains just something about MTGA's network settings. I am writing some recap of what I have done, and hopefully I can make something playable or hackable out of it.

P.S. I've notice some duplicate in data transfer, where some inconsistency (a.k.a. bug) might lie. But I'm somewhat afraid to exploit it. Did you create a new account to test your exploit or just risk being banned? :P
